### PR TITLE
Optimize bitreverse codegen

### PR DESCRIPTION
**Pull request by arpankapoor**
Hi, this is my attempt at bringing the bitreverse codegen on par with the LLVM backend (atleast for x86_64).
The algorithm is from [go's math library](https://cs.opensource.google/go/go/+/refs/tags/go1.20.2:src/math/bits/bits.go;l=253).
The 8/16 bit codegen is similar to the earlier version, while the 32/64 bit codegen is ~30% shorter (for x86_64).

Following are the 64-bit bitreverse codegen (for x86_64) before and after the changes:
before:
```assembly
movabs  rax, -140733193420800
rorx    rdi, rdi, 32
mov rdx, rdi
shr rdi, 17
movabs  rsi, -143978844372567040
sal rdx, 15
and rdx, rax
movabs  rax, 140733193420799
and rdi, rax
or  rdx, rdi
mov rcx, rdx
shr rcx, 10
xor rcx, rdx
mov rax, rcx
sal rax, 10
and rax, rsi
movabs  rsi, 17873794169274399
and rcx, rsi
movabs  rsi, -2290007770269203952
or  rax, rcx
xor rax, rdx
mov rcx, rax
shr rcx, 4
xor rcx, rax
mov rdx, rcx
sal rdx, 4
and rdx, rsi
movabs  rsi, 1009796018965021729
and rcx, rsi
movabs  rsi, -8565247220761091832
or  rdx, rcx
xor rdx, rax
mov rcx, rdx
shr rcx, 2
xor rcx, rdx
lea rax, 0[0+rcx*4]
and rax, rsi
movabs  rsi, 2470374213237114946
and rcx, rsi
or  rax, rcx
xor rax, rdx
ret
```
after:
```assembly
movabs  rdx, 6148914691236517205
mov rax, rdi
shr rdi
movabs  rcx, 1085102592571150095
and rdi, rdx
add rax, rax
movabs  rdx, -6148914691236517206
and rax, rdx
or  rdi, rax
movabs  rax, 3689348814741910323
mov rdx, rdi
sal rdi, 2
shr rdx, 2
and rdx, rax
movabs  rax, -3689348814741910324
and rdi, rax
or  rdx, rdi
mov rax, rdx
sal rdx, 4
shr rax, 4
and rax, rcx
movabs  rcx, -1085102592571150096
and rdx, rcx
or  rax, rdx
bswap   rax
ret
```

Godbolt link comparing LLVM and the current codegen: https://godbolt.org/z/43916e3sj
Partially addresses #144.

### CODE REVIEW

File: example/mini_core_hello_world.rs [hunk 1]
```diff
@@ -168,6 +168,11 @@ fn main() {
         world as Box<dyn SomeTrait>;
 
         assert_eq!(intrinsics::bitreverse(0b10101000u8), 0b00010101u8);
+        assert_eq!(intrinsics::bitreverse(0xddccu16), 0x33bbu16);
+        assert_eq!(intrinsics::bitreverse(0xffee_ddccu32), 0x33bb77ffu32);
+        assert_eq!(intrinsics::bitreverse(0x1234_5678_ffee_ddccu64), 0x33bb77ff1e6a2c48u64);
+        // != cannot be applied to type u128?
```
Comment (antoyo) [inline example/mini_core_hello_world.rs:174]: You probably need to add an `impl PartialEq for u128` in `example/mini_core.rs`.

**Review by antoyo — COMMENTED** [review]
Thank you!

File: example/mini_core_hello_world.rs (hunk 1 above)
Comment (arpankapoor) [reply example/mini_core_hello_world.rs]: That fixes it, thanks.

**Comment by antoyo** [comment]
It seems you might have found a bug. The [CI without int128](https://github.com/rust-lang/rustc_codegen_gcc/actions/runs/4523501235/jobs/7966670882?pr=257#step:19:56) broke.

I'll see if I can find a quick fix.

**Comment by antoyo** [comment]
Ah no. It's "normal": `#[no_core]` means it won't link the compiler-rt stuff. I opened [an issue](https://github.com/rust-lang/rustc_codegen_gcc/issues/258) for this.
So, for now, could you please comment this line?

**Comment by antoyo** [comment]
Actually, you need to move the assert to `std_example`.

**Comment by antoyo** [comment]
> I've moved the assert. Are the failing CI tests still related to this change?

No, the `--test-failing-rustc` are supposed to fail and the other failure is [this issue](https://github.com/rust-lang/rustc_codegen_gcc/issues/245).

I'll review and merge later today.

**Comment by antoyo** [comment]
Thanks a lot!

**Pull request merged**